pkg/fanal/analyzer/language/nodejs/yarn/yarn.go

```diff
@@ -177,3 +210,148 @@ func (a yarnAnalyzer) parsePackageJsonDependencies(fsys fs.FS, path string) (map
+	dir, fileName := filepath.Split(path)
+	dir = strings.TrimRight(dir, "/")
+	dirs = strings.Split(dir, "/")
+	return dirs, fileName
+}
+
+func (a yarnAnalyzer) traverseYarnClassicPkgs(fsys fs.FS, path string, fn traverseFunc) error {
+	// Traverse node_modules dir
+	// Note that fs.FS is always slashed regardless of the platform,
+	// and path.Join should be used rather than filepath.Join.
+	err := fsutils.WalkDir(fsys, path, isNodeModulesPkg, func(filePath string, d fs.DirEntry, r dio.ReadSeekerAt) error {
+		pkg, err := a.packageJsonParser.Parse(r)
+		if err != nil {
+			return xerrors.Errorf("unable to parse %q: %w", filePath, err)
+		}
+		return fn(pkg)
+	})
+	if err != nil {
+		return xerrors.Errorf("walk error: %w", err)
+	}
+	return nil
+}
+
+func (a yarnAnalyzer) traverseYarnModernPkgs(fsys fs.FS, root string, fn traverseFunc) error {
+	yarnDir := path.Join(root, ".yarn")
+	if _, err := fs.Stat(fsys, yarnDir); errors.Is(err, fs.ErrNotExist) {
+		return nil
+	} else if err != nil {
+		return xerrors.Errorf("unable to parse %q: %w", yarnDir, err)
+	}
+
+	if err := a.traverseUnpluggedFolder(fsys, yarnDir, fn); err != nil {
+		return err
+	}
+	if err := a.traverseCacheFolder(fsys, yarnDir, fn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a yarnAnalyzer) traverseUnpluggedFolder(fsys fs.FS, root string, fn traverseFunc) error {
+	// `unplugged` hold machine-specific build artifacts
+	unpluggedPath := path.Join(root, "unplugged")
+	if _, err := fs.Stat(fsys, unpluggedPath); err != nil {
+		return nil
+	}
+
+	// Traverse .yarn/unplugged dir
+	err := fsutils.WalkDir(fsys, unpluggedPath, isNodeModulesPkg, func(path string, d fs.DirEntry, r dio.ReadSeekerAt) error {
+		pkg, err := a.packageJsonParser.Parse(r)
+		if err != nil {
+			return xerrors.Errorf("unable to parse %q: %w", path, err)
+		}
+		return fn(pkg)
+	})
+	if err != nil {
+		return xerrors.Errorf("walk error: %w", err)
+	}
+
+	return nil
+}
+
+func (a yarnAnalyzer) traverseCacheFolder(fsys fs.FS, root string, fn traverseFunc) error {
+	cachePath := path.Join(root, "cache")
+	if _, err := fs.Stat(fsys, cachePath); err != nil {
+		return nil
+	}
+
+	required := func(path string, _ fs.DirEntry) bool {
+		return filepath.Ext(path) == ".zip"
+	}
+
+	// Traverse .yarn/cache dir
+	err := fsutils.WalkDir(fsys, cachePath, required, func(path string, d fs.DirEntry, r dio.ReadSeekerAt) error {
+		fi, err := d.Info()
+		if err != nil {
+			return xerrors.Errorf("file stat error: %w", err)
+		}
+
+		zr, err := zip.NewReader(r, fi.Size())
+		if err != nil {
+			return xerrors.Errorf("zip reader error: %w", err)
+		}
+
+		for _, f := range zr.File {
+			if filepath.Base(f.Name) != types.NpmPkg {
+				continue
+			}
+			pkgFile, err := f.Open()
+			if err != nil {
+				return xerrors.Errorf("file open error: %w", err)
+			}
+			defer pkgFile.Close()
+			pkg, err := a.packageJsonParser.Parse(pkgFile)
+			if err != nil {
+				return xerrors.Errorf("unable to parse %q: %w", path, err)
+			}
```

Comment:
You are using `defer` in a loop. Can we move this logic into a separate function to close each file?

Reply:
Great catch! Done.

Reply:
We use new function to only parse `package.json` **from zip file**. The name `parsePackageJson` can be confusing (I mean you might think that this function for any `package.json`).
What if we use `parsePackageJsonFromZip` name?

Reply:
done